java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java

```diff
@@ -0,0 +1,241 @@
+import org.netbeans.spi.editor.hints.Fix;
+import org.netbeans.spi.editor.hints.Severity;
+import org.openide.text.Line;
+import org.openide.util.NbBundle;
+
+import static org.netbeans.modules.maven.hints.pom.Bundle.*;
+
+/**
+ * Converts source/target javac options to the release option where possible (JEP 247).
+ * @author mbien
+ */
+@NbBundle.Messages({
+    "TIT_UseReleaseVersionHint=Convert matching source/target javac options to the release option.",
+    "DESC_UseReleaseVersionHint=Matching source/target options can be replaced with a single release option (JEP 247). This enforces API compatibility with the chosen Java release.",
+    "FIX_UseReleaseVersionHint=Convert to release option for strict compatibility checks."})
+public class UseReleaseOptionHint implements POMErrorFixProvider {
+
+    private static final String TARGET_TAG = "target";
+    private static final String SOURCE_TAG = "source";
+    private static final String RELEASE_TAG = "release";
+
+    private static final Configuration config = new Configuration(UseReleaseOptionHint.class.getName(),
+                TIT_UseReleaseVersionHint(), DESC_UseReleaseVersionHint(), true, Configuration.HintSeverity.WARNING);
+
+    @Override
+    public List<ErrorDescription> getErrorsForDocument(POMModel model, Project prj) {
+
+        List<ErrorDescription> hints = new ArrayList<>();
+
+        Build build = model.getProject().getBuild();
+        if (build != null) {
+            for (Plugin plugin : build.getPlugins()) {
+                if ("maven-compiler-plugin".equals(plugin.getArtifactId())) {
+                    if (!isPluginCompatible(plugin)) {
+                        return Collections.emptyList();
+                    }
+                    if (plugin.getConfiguration() != null) {
+                        hints.addAll(createHintsForParent("", plugin.getConfiguration()));
+                    }
+                    if (plugin.getExecutions() != null) {
+                        for (PluginExecution execution : plugin.getExecutions()) {
+                            if (execution.getConfiguration() != null) {
+                                hints.addAll(createHintsForParent("", execution.getConfiguration()));
+                            }
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+
+        Properties properties = model.getProject().getProperties();
+        if (properties != null) {
+            hints.addAll(createHintsForParent("maven.compiler.", properties));
+        }
+
+        return hints;
+    }
+
+    private List<ErrorDescription> createHintsForParent(String prefix, POMComponent parent) {
+
+        int source;
+        int target;
+
+        // property name or an int value
+        String release = null;
+
+        try {
+            String sourceText = parent.getChildElementText(POMQName.createQName(prefix+SOURCE_TAG, true));
+            if (isProperty(sourceText)) {
```

Comment:
the hint will only show up if there the pom has both, source and target defined and both are set to the same value. It should simply give up if one of the two is not valid or missing.

The main motivation behind this hint was because I met many who still didn't know that the release option exists and why it is better. I personally never set one without the other. Haven't thought about converting a single option into release - you think this would be save in all cases?

Reply:
No, as it only triggers om same strings, it is functionally OK as it is, but the code does not express that intention much clearly :) Maybe an explanation comment that a catch-all-errors handler is OK would be helpful.

Reply:
ah i see. Adding a comment to the catch block.